src/component.js

```diff
@@ -0,0 +1,572 @@
+    }
+  }
+
+  /*
+  * Find entity inside children or pop up to check parent and continue search there
+  * params:
+  */
+  async find(ids, recursive = true, depth = 0, exclude = []) {
+    let component = null;
+    //console.log(this.id, ' ', ids);
+    if (ids.length) {
+      const id = ids[0];
+      let nIds = ids;
+      let nRecursive = true;
+      //
+      if (this.id === id) {
+        component = this;
+        if (ids.length > 1) {
+          nIds = ids.slice(1);
+          nRecursive = false;
+        }
+      }
+      // recursive search
+      if ((component && (ids.length > 1)) || (!component && recursive)) {
+        for (const item of await this.getIDs()) {
+          if (!exclude.includes(item)) {
+            const c = await this.buildChild(item, false);
+            if (c) {
+              component = await c.find(nIds, nRecursive, depth + 1);
+              if (component) {
+                break;
+              }
+            }
+          }
+        }
+      }
+      // try to use one level upper
+      if (!component && this.parent && (depth === 0)) {
+        component = this.parent.find(ids, true, 0, [this.id]);
+      }
+    }
+    return component;
+  }
+
+  async resolve(components, resolveEmptyToThis = false) {
+    let r = [];
+    if (components.length) {
+      for (let component of components) {
+        // find component
+        const c = await this.find(component.split('/'));
+        if (c) {
+          r.push(c);
+        } else {
+          this.logger.error(`Component '${component}' was not found`);
+        }
+      }
+      return r;
+    } else {
+      if (resolveEmptyToThis) {
+        r.push(this);
+      }
+    }
+    return r;
+  }
+
+  async unfoldHierarchy(anchor, id, home, unique = true, list = []) {
+    if (unique) {
+      list = list.filter((item) => (item.anchor !== anchor) || (item.component.uuid !== this.uuid) && (item.anchor === anchor));
+    }
+    list.push({ component: this, anchor, id, home });
+
+    // check depends list first
+    let depends = [];
+    for (const desc of this.descriptions) {
+      if (desc.depends) {
+        depends = depends.concat(await desc.depends(this.tln));
+      }
+    }
+    for (let component of await this.resolve(depends)) {
+      list = await component.unfoldHierarchy(component.uuid, component.id, component.home, unique, list);
+    }
+
+    // check inherits list first
+    let inherits = [];
+    for (const desc of this.descriptions) {
+      if (desc.inherits) {
+        inherits = inherits.concat(await desc.inherits(this.tln));
+      }
+    }
+    for (let component of await this.resolve(inherits)) {
+      list = await component.unfoldHierarchy(anchor, id, home, unique, list);
+    }
+    // check parents' hierarchy
+    if (this.parent) {
+      list = await this.parent.unfoldHierarchy(anchor, id, home, unique, list);
+    }
+    return list;
+  }
+
+  static getDependsList(hierarchy, anchor) {
+    const r = [];
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of the unused 'r' variable. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://159.65.230.204:9000/coding_rules#rule_key=javascript%3AUnusedVariable)